README.md
# Pouchdb-Paginated-Query

`Pouchdb-Paginated-Query` performs paginated PouchDB queries, and provides each result via an async iterable. It works well with [await-for-it](https://github.com/toolbuilder/await-for-it), which joins async iterables with event handling, transformations, and pub/sub.

Here is a very simple example.

```javascript
import { allRows } from '@toolbuilder/pouchdb-paginated-query'

const db = new PouchDB('db', { ..your options here.. })
// allRows interprets 'limit' as the page size
const paginatedQuery = allRows(db, { startkey: '03', endkey: '16', limit: 5 })
for await (const row of paginatedQuery) {
  // handle each returned row here
}
```

Here is the simplest example with [await-for-it](https://github.com/toolbuilder/await-for-it).

```javascript
import { allRows } from '@toolbuilder/pouchdb-paginated-query'
import { chainable } from '@toolbuilder/await-for-it'

const db = new PouchDB('db', { ..your options here.. })
// allDocs interprets 'limit' as the page size
const paginatedQuery = allRows(db, { startkey: '03', endkey: '16', limit: 5, include_docs: true })
chainable(paginatedQuery)
  .map(row => row.doc)
  .callAwait(async doc => /* your processing function here */)
  .catch(error => { /* your exception handler here */ })
  .finally(() => { /* your cleanup function here */ })
  .run() // start asynchronous processing

```

## Installation

```bash
npm install --save @toolbuilder/pouchdb-paginated-query
```

## API

* [allDocs](#allDocs)
* [allRows](#allRows)
* [queries](#queries)

### **allDocs**

Paginated PouchDB query responses provided as an async iterator. If pagination is selected as an option, responses are provided until the response contains no entries or an error is thrown.

By default, pagination is performed if the `limit` field has a value. To turn it off, set options.paginate = false.

Pagination works by updating the 'startkey' field of the options parameter to query for the next batch of rows when 'options.limit' is less than the number
of records specified by the query. It also updates the 'skip' field after the first page to prevent record skipping or duplicates. See <https://pouchdb.com/2014/04/14/pagination-strategies-with-pouchdb.html> for more information on paginated queries.

* `db` **PouchDB** An open PouchDB database. In case you wish to proxy the database for some reason, only `db.allDocs` will be called by this generator.
* `options` **Object** Any valid `PouchDB.allDocs(options)` query parameter. For pagination to work you should at least set the 'limit' field. The startkey, and 'skip' option parameters will be changed as required by this generator to manage pagination. The query will NOT be paginated if `options.limit == null || options.paginate === false`

Throws **PouchError** if any exception is thrown by PouchDB.

Returns **AsyncIterable** where each value is an unmodified PouchDB.allDocs() return value.

```javascript
import PouchDB from 'pouchdb'
import { allDocs } from '@toolbuilder/pouchdb-paginated-query'

const db = new PouchDB('db', { ..your options here.. })
const paginatedQuery = allDocs(db, { startkey: '03', endkey: '16', limit: 5 })
for await (const response of paginatedQuery) {
  const rows = response.total_rows
  for (const row of response.rows) {
    // handle rows here
  }
}
```

### **allRows**

The rows from paginated queries provided as an async iterator. If pagination is selected as an option, rows are provided until the PouchDB response contains no entries or an error is thrown.

This function is a convenience wrapper around [allDocs](#allDocs) that returns just the rows.

* `db` **PouchDB** An open PouchDB database. In case you wish to proxy the database for some reason, only `db.allDocs` will be called by this generator.
* `options` **Object** Any valid `PouchDB.allDocs(options)` query parameter. For pagination to work you should at least set the 'limit' field. The startkey, and 'skip' option parameters will be changed as required by this generator to manage pagination. The query will NOT be paginated if `options.limit == null || options.paginate === false`.

Throws **PouchError** if any exception is thrown by PouchDB.

Returns **AsyncIterable** where each value is a row object from a PouchDB query.

```javascript
import PouchDB from 'pouchdb'
import { allDocs } from '@toolbuilder/pouchdb-paginated-query'

const db = new PouchDB('db', { ..your options here.. })
const paginatedQuery = allRows(db, { startkey: '03', endkey: '16', limit: 5 })
for await (const response of paginatedQuery) {
  const id = response.id
  const doc = response.doc
  // row processing here
}
```

### **queries**

Iterate over a sequence of queries, using allDocs, allRows, or your own custom function as paginatedQueryFn.

* `db` **PouchDB** An open PouchDB database. In case you wish to proxy the database for some reason, only `db.allDocs` will be called by this generator.
* `paginatedQueryFn` **AsyncGeneratorFunction** any function that matches the signature of allDocs or allRows. The function will be called once for each query provided.

Throws **PouchError** if any exception is thrown by PouchDB.

Returns **AsyncIterable** where each value is yielded from a `paginatedQueryFn` iterable.

```javascript
import PouchDB from 'pouchdb'
import { queries, allRows } from '@toolbuilder/paginated-pouchdb-query'

const db = new PouchDB('db', { ..your options here.. })
const someQueries = [{ /* query options */ }, { /* query options */ }]
// Call allRows for each query and process the results
for await (const row of queries(db, allRows)(someQueries)) {
  // process rows from both queries here
}
```

This example uses [await-for-it](https://github.com/toolbuilder/await-for-it) to query PouchDB in response to events.

```javascript
import { chainable, Queue } from '@toolbuilder/await-for-it'
import { RingBuffer } from '@toolbuilder/ring-buffer'
import PouchDB from 'pouchdb'
import { queries, allRows } from '@toolbuilder/paginated-pouchdb-query'

// Setup queue to keep only the most recent 100 data values.
// Older values that have not been processed yet are quietly dropped.
// This behavior is provided by RingBuffer. Other buffers behave differently.
// For example, an Array will never drop anything, but could grow forever.
const queue = new Queue(new RingBuffer(100))

// setup an async iterator to handle the input events
const controller = chainable(queue)
  .filter(event => dropEventsThatYouDoNotWant(event))
  .map(async (event) => generateQueryOptionsFromEvent(event))
  // Perform query, and return rows for each event
  .mapWith(queries(db, allRows))
  .map(async (row) => processRow(row))
  // An error stops all iteration
  .catch(error => handleErrors(error))
  .finally(async () => doSomeCleanup())
  .run() // you could use 'publish' here if you want pub/sub

// In your event handler, push events into the queue
const eventHandler = (event) => queue.push(event)
// register your event handler of course

// You can stop/resume the chainable iterator if you need to.
// You can still push to the queue while chainable is stopped,
// but the buffer might drop what you push depending on the buffer
// you choose.
if (controller.running) controller.stop()
controller.start()

// Somewhere else, such as an 'application done' event, tell the
// Queue that it is done so that the chainable finally method is called
// By default all values in the buffer are processed before iteration stops.
queue.done()

// Elsewhere, you can also pass an exception to the queue, so that the async
// iterator will handle it asynchronously.
queue.exception(new YourError(someData))
// The .catch() method in the chainable above will now get
// YourError, and iteration will stop

```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.3](https://github.com/toolbuilder/pouchdb-paginated-query/compare/v0.1.2...v0.1.3) (2022-01-26)

### [0.1.2](https://github.com/toolbuilder/pouchdb-paginated-query/compare/v0.1.1...v0.1.2) (2022-01-05)


### Features

* dual package - ES module with CommonJS support and Electron tests ([e4cf18a](https://github.com/toolbuilder/pouchdb-paginated-query/commit/e4cf18a0b7e31e37538b589d5ae9bb7b6364d212))

### 0.1.1 (2020-11-28)


### Features

* modified API, created documentation ([ce22923](https://github.com/toolbuilder/pouchdb-paginated-query/commit/ce22923e7d3769d6869d7ea9b53bc3252c6c10f9))
* pagination working and tested, a few corner cases to go ([171caba](https://github.com/toolbuilder/pouchdb-paginated-query/commit/171caba662b4adfac81438b8ab87365b2f386d48))


rollup.config.test.js
import { baseBrowserTestConfig, nodeConfigs } from '@toolbuilder/rollup-plugin-test-tools'

const options = {
  packageName: 'pouchdb-paginated-query-browser',
  input: ['test/**/*test.js'], // unit test source file glob
  aliasOptions: { // 1: select different file to import than Node pack file test uses
    entries: [
      { find: './database.node.js', replacement: './database.browser.js' }
    ]
  }, // file aliases for browser build
  external: ['pouchdb'], // 3a: Make PouchDB external
  globals: {
    pouchdb: 'PouchDB' // 3b: and specify global name exported by PouchDB UMD package
  },
  checkSemverConflicts: true, // check for semver range conflicts between testPackageJson and packageJson
  testPackageJson: {
    type: 'module', // whether module or commonjs doesn't seem to matter for the test.
    scripts: { // 2: test script changed to concatenate the PouchDB distribution files for browser with the test UMD file
      test: 'cat ./node_modules/pouchdb/dist/pouchdb.min.js ./node_modules/pouchdb/dist/pouchdb.memory.min.js all.umd.js | tape-run'
    },
    // Auto-magically adds external dependencies calculated
    // by relativeToPackage, so no need to list them.
    devDependencies: {
    // dependencies for test script
      'cash-cat': '^0.2.0', // provides 'cat' for non-POSIX shells
      'tape-run': '^9.0.0'
    }
  }
}

export default [...nodeConfigs, ...baseBrowserTestConfig(options)]
// export default [...browserTestConfig]


test/database.node.js
import PouchDB from 'pouchdb'
import InMemoryAdapter from 'pouchdb-adapter-memory'

export const Database = PouchDB.plugin(InMemoryAdapter)


test/paginator.test.js
import { test } from 'zora'
import { chainable } from '@toolbuilder/await-for-it'
import { allRows, allDocs, queries } from '../src/paginator.js'
import { Database } from './database.node.js'

const uniqueIdGenerator = () => {
  let number = 0
  return () => { number++; return number.toString().padStart(2, '0') }
}

const uniqueId = uniqueIdGenerator()

// Create some ids for the test records, and for testing.
const createTestIds = async () => chainable
  .range(20)
  .map(id => id.toString().padStart(2, '0'))
  .toArray()

// Create, and insert test records into the DB.
const insertTestRecords = async (db, testIds) => {
  const records = await chainable(testIds)
    // zero pad id so JS string ordering provides expected results
    .map(id => id.toString().padStart(2, '0'))
    .map(_id => ({ _id, text: `this is ${_id}` }))
    .toArray()
  return db.bulkDocs(records)
}

const newDbWithTestRecords = async (testIds) => {
  const dbId = `db-${uniqueId()}`
  const db = new Database(dbId, { adapter: 'memory' })
  await insertTestRecords(db, testIds)
  return db
}

const paginated = true
test('proper number of records returned for each query', async assert => {
  const testIds = await createTestIds()
  const testCases = [
    [
      'ascending paginated query made expected number of queries',
      testIds.slice(3, 17),
      {
        startkey: '03',
        endkey: '16',
        limit: 5,
        include_docs: true,
        descending: false
      },
      paginated
    ],
    [
      'options.paginate == true with limit value causes pagination',
      testIds.slice(3, 17),
      {
        startkey: '03',
        endkey: '16',
        limit: 5,
        include_docs: true,
        descending: false,
        paginate: true // if limit not provided, this makes no sense
      },
      paginated
    ],
    [
      'descending paginated query made expected number of queries',
      testIds.slice(3, 17).reverse(),
      {
        startkey: '16',
        endkey: '03',
        limit: 5,
        include_docs: true,
        descending: true
      },
      paginated
    ],
    [
      'paginated query returns no rows',
      [],
      {
        startkey: '458',
        endkey: '459',
        limit: 5,
        include_docs: true,
        descending: false
      },
      paginated
    ],
    [
      'non-paginated query because no limit specified',
      testIds.slice(3, 17),
      {
        startkey: '03',
        endkey: '16',
        include_docs: true,
        descending: false
      },
      !paginated
    ],
    [
      'paginate option stops pagination',
      testIds.slice(3, 8),
      {
        startkey: '03',
        endkey: '16',
        limit: 5,
        include_docs: true,
        descending: false,
        paginate: false
      },
      !paginated
    ]
  ]

  // Run testCases
  await chainable(testCases)
    .forEach(async ([description, expectedIds, query, isPaginated]) => {
      const db = await newDbWithTestRecords(testIds)
      const actualRowCounts = await chainable(allDocs(db, query))
        .map(response => response.rows.length)
        .catch(error => console.log(`Query count Pagination error: ${error}`))
        .finally(async () => db.close())
        .toArray()

      let expectedRowCounts = [expectedIds.length] // for non-paginated queries
      if (isPaginated) {
        // Simulate query (with slice), and pagination (with chunk), to calculate expected row counts
        expectedRowCounts = await chainable(expectedIds).chunk(query.limit, 1).map(chunk => chunk.length).toArray()
        expectedRowCounts.push(0) // to account for last empty query of pagination
      }

      assert.deepEqual(actualRowCounts, expectedRowCounts, description)
    })
})

test('rows', async assert => {
  const testIds = await createTestIds()
  const testCases = [
    [
      'all expected ids returned',
      testIds.slice(3, 17),
      {
        startkey: '03',
        endkey: '16',
        limit: 5,
        include_docs: false,
        descending: false
      }
    ],
    [
      'query response with no rows handled',
      [],
      {
        startkey: '20',
        endkey: '36',
        limit: 5,
        include_docs: false,
        descending: false
      }
    ],
    [
      'non-paginated query handled',
      testIds.slice(3, 17),
      {
        startkey: '03',
        endkey: '16',
        // limit: 5 // removing this field eliminates pagination
        include_docs: false,
        descending: false
      }
    ],
    [
      'missing startkey handled',
      testIds,
      {
        limit: 5,
        include_docs: true,
        descending: false
      }
    ],
    [
      'initial skip supported',
      testIds.slice(3),
      {
        limit: 2,
        skip: 3,
        include_docs: true,
        descending: false
      }
    ],
    [
      'include_docs === false supported',
      testIds.slice(3),
      {
        limit: 2,
        skip: 3,
        include_docs: false,
        descending: false
      }
    ],
    [
      'limit zero returns nothing',
      [],
      {
        limit: 0
      }
    ],
    [
      'descending true works correctly',
      testIds.slice(3, 17).reverse(),
      {
        startkey: '16',
        endkey: '03',
        limit: 5,
        include_docs: false,
        descending: true
      }
    ]
  ]

  // TODO: test with options.key and options.keys

  // Run testCases
  await chainable(testCases)
    .forEach(async ([description, expectedIds, query]) => {
      const db = await newDbWithTestRecords(testIds)

      const actualIds = await chainable(allRows(db, query))
        .map(row => row.id)
        .catch(error => console.log(`Pagination error: ${error}`))
        .finally(async () => db.close())
        .toArray()

      assert.deepEqual(actualIds, expectedIds, description)
    })
})

test('query stream', async assert => {
  const testIds = await createTestIds()
  const db = await newDbWithTestRecords(testIds)

  const queryStream = [ // three queries
    { startkey: '03', endkey: '16', limit: 5, include_docs: true, descending: false },
    { startkey: '03', endkey: '06', limit: 5, include_docs: false, descending: false },
    { startkey: '07', endkey: '16', limit: 5, include_docs: true, descending: false }
  ]

  const actualIds = await chainable(queryStream)
    .mapWith(queries(db, allRows))
    .map(row => row.id)
    .catch(error => console.log(`Pagination error: ${error}`))
    .finally(() => db.close())
    .toArray()

  // Simulate queries to get expected results
  const expectedIds = await chainable(queryStream)
    .map(query => [Number(query.startkey), Number(query.endkey) + 1])
    .map(range => testIds.slice(...range))
    .flatten()
    .toArray()

  assert.deepEqual(actualIds, expectedIds, 'all output from multiple queries as expected')
})


src/paginator.js
const flatten = async function * (iterable) {
  for await (const value of iterable) {
    yield * value
  }
}

const map = async function * (fn, iterable) {
  for await (const value of iterable) {
    yield fn(value)
  }
}

/**
 * Paginated query generator function that produces a sequence of query responses
 * until the response contains no entries or an error is thrown.
 *
 * Pagination works by updating the 'startkey' field of the options parameter to
 * query for the next batch of rows when 'options.limit' is less than the number
 * of records specified by the query. It also updates the 'skip' field after the
 * first page to prevent record skipping or duplicates.
 *
 * See https://pouchdb.com/2014/04/14/pagination-strategies-with-pouchdb.html for
 * more information on paginated queries.
 *
 * @param {PouchDB} db - An open PouchDB database. In case you wish to proxy the
 * database for some reason, only db.allDocs will be called by this generator.
 * @param {Object} options - Any valid PouchDB.allDocs(options) query parameter. For
 * pagination to work you should at least set the 'limit' field. The 'startkey, and
 * 'skip' option parameters will be changed as required by this generator to manage
 * pagination.
 * @throws error {PouchError} - any exception thrown by PouchDB
 * @example
 * const db = new PouchDB('db', { ..your options here.. })
 * const paginatedQuery = allDocs(db, { startkey: '03', endkey: '16', limit: 5 })
 * for (const response of paginatedQuery) {
 *  // handle each query response here
 * }
 */
const paginatedQuery = async function * (db, options) {
  const queryOptions = { ...options }
  let response = await db.allDocs(queryOptions)
  // No need to paginate for options.key,
  // and startkey is not allowed for options.keys
  if (options.key != null || options.keys != null) return response
  while (response && response.rows.length > 0) {
    // this is how the PouchDb team recommends doing paginated queries
    // pick up the last key returned to use as starting point for next query
    queryOptions.startkey = response.rows[response.rows.length - 1].id
    // skip the next starting point, since it's in this response
    queryOptions.skip = 1

    yield response // so client can access the data
    response = await db.allDocs(queryOptions)
  }
  yield response // the final response, which contains no rows
}

const nonPaginatedQuery = async function * (db, options) {
  yield db.allDocs(options)
}

/**
 * PouchDB query response provided as an async iterator. If pagination is selected
 * as an option, responses are provided until the response contains no entries or
 * an error is thrown. Only responses containing rows are returned.
 *
 * By default, pagination is on if the limit field has a value. To turn it off,
 * set options.paginate = false.
 *
 * Pagination works by updating the 'startkey' field of the options parameter to
 * query for the next batch of rows when 'options.limit' is less than the number
 * of records specified by the query. It also updates the 'skip' field after the
 * first page to prevent record skipping or duplicates.
 *
 * See https://pouchdb.com/2014/04/14/pagination-strategies-with-pouchdb.html for
 * more information on paginated queries.
 *
 * @param {PouchDB} db - An open PouchDB database. In case you wish to proxy the
 * database for some reason, only db.allDocs will be called by this generator.
 * @param {Object} options - Any valid PouchDB.allDocs(options) query parameter. For
 * pagination to work you should at least set the 'limit' field. The 'startkey, and
 * 'skip' option parameters will be changed as required by this generator to manage
 * pagination. The query will NOT be paginated if options.limit == null || options.paginate === false
 * @throws error {PouchError} - any exception thrown by PouchDB
 * @example
 * const db = new PouchDB('db', { ..your options here.. })
 * const paginatedQuery = allDocs(db, { startkey: '03', endkey: '16', limit: 5 })
 * for await (const response of paginatedQuery) {
 *  // handle each query response here
 * }
 */
export const allDocs = (db, options) => {
  const nonPaginated = options.limit == null || options.paginate === false
  return nonPaginated ? nonPaginatedQuery(db, options) : paginatedQuery(db, options)
}

/**
 * The rows from paginated queries provided as an async iterator. If pagination is selected
 * as an option, rows are provided until the PouchDB response contains no entries or
 * an error is thrown.
 *
 * This function is a convenience wrapper around allDocs that returns just the rows.
 *
 * @param {PouchDB} db - An open PouchDB database. Only 'db.allDocs' is called in case
 * you wish to proxy the database for some reason such as special error handling.
 * @param {Object} options - Any valid PouchDB.allDocs(options) query parameter. The query
 * will NOT be paginated if options.limit == null || options.paginate === false
 * @throws error {PouchError} - any exception thrown by PouchDB
 * @example
 * const db = new PouchDB('db', { ..your options here.. })
 * const paginatedQuery = rows(db, { startkey: '03', endkey: '16', limit: 5, include_docs: true })
 * for await (const row of paginatedQuery) {
 *  console.log(row.doc._id)
 * }
 */
export const allRows = (db, options) => flatten(map(response => response.rows, allDocs(db, options)))

/**
 * Iterate over a sequence of queries, using either allDocs or rows as paginatedQueryFn.
 *
 * @param {PouchDB} db - An open PouchDB database. Only 'db.allDocs' is called in case
 * you wish to proxy the database for some reason such as special error handling.
 * @param {AsyncGeneratorFunction} paginatedQueryFn - any function that matches the signature of allDocs or allRows.
 */
export const queries = (db, paginatedQueryFn) => (queries) => flatten(map(query => paginatedQueryFn(db, query), queries))
